nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs

```diff
@@ -295,9 +304,218 @@ impl<T: Hash + Eq> Iterator for AvailableIterator<'_, T> {
+        match external_ip {
+            OmicronZoneExternalIp::Floating(ip) => {
+                let ip = ip.ip;
+                if self.used_snat_ips.contains_key(&ip) {
+                    bail!(
+                        "mismatched IP: {external_ip:?} also used as SNAT IP"
+                    );
+                }
+                Ok(self.used_exclusive_ips.contains(&ip))
+            }
+            OmicronZoneExternalIp::Snat(snat) => {
+                let ip = snat.snat_cfg.ip;
+                let port_range =
+                    SnatPortRange::try_from(snat.snat_cfg.port_range_raw())?;
+
+                if self.used_exclusive_ips.contains(&ip) {
+                    bail!(
+                        "mismatched IP: {external_ip:?} also used as floating"
+                    );
+                }
+
+                Ok(self
+                    .used_snat_ips
+                    .get(&ip)
+                    .map(|port_ranges| port_ranges.contains(&port_range))
+                    .unwrap_or(false))
+            }
+        }
+    }
+
+    fn claim_next_exclusive_ip(&mut self) -> Result<IpAddr, Error> {
+        for ip in &mut *self.service_ip_pool_ips {
+            if !self.used_snat_ips.contains_key(&ip)
+                && self.used_exclusive_ips.insert(ip)
+            {
+                return Ok(ip);
+            }
+        }
+
+        Err(Error::NoExternalServiceIpAvailable)
+    }
+
+    // This is currently only used by a unit test, but will be used by real code
+    // once we support boundary NTP zone allocation.
+    #[cfg(test)]
+    fn claim_next_snat_ip(&mut self) -> Result<SourceNatConfig, Error> {
+        // Prefer reusing an existing SNAT IP, if we still have port ranges
+        // available on that ip.
+        for (ip, used_port_ranges) in self.used_snat_ips.iter_mut() {
+            for port_range in SnatPortRange::iter() {
+                if used_port_ranges.insert(port_range) {
+                    return Ok(port_range.into_source_nat_config(*ip));
+                }
+            }
+        }
+
+        // No available port ranges left; allocate a new IP.
+        for ip in &mut *self.service_ip_pool_ips {
+            if self.used_exclusive_ips.contains(&ip) {
+                continue;
+            }
+            match self.used_snat_ips.entry(ip) {
+                // We already checked all occupied slots above; move on.
+                Entry::Occupied(_) => continue,
+                Entry::Vacant(slot) => {
+                    let port_range = SnatPortRange::One;
+                    slot.insert(BTreeSet::new()).insert(port_range);
+                    return Ok(port_range.into_source_nat_config(ip));
+                }
+            }
+        }
+
+        Err(Error::NoExternalServiceIpAvailable)
+    }
+}
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, strum::EnumIter,
+)]
+#[cfg_attr(test, derive(test_strategy::Arbitrary))]
+enum SnatPortRange {
```

Comment:
Oh hey, this is what we talked about at some point right?

Reply:
Almost - we talked about it in the bigger context of `omicron-common`; this is more of a local trial run of that idea. I think I like it, but it's not entirely frictionless.